### PR TITLE
Add option to not reveal the instant or preliminary scores

### PR DESCRIPTION
**Pull request by peterdietz**
Add an admin-level option to choose to show or hide the preliminary scores. Also, updates the cache length time to be based upon preliminary score, vs 60 minute delay.

### CODE REVIEW

File: db/install.xml [hunk 1]
```diff
@@ -9,13 +9,14 @@
...
         <FIELD NAME="cm" TYPE="int" LENGTH="10" NOTNULL="true" UNSIGNED="true" DEFAULT="0" SEQUENCE="false" PREVIOUS="id" NEXT="userid"/>
         <FIELD NAME="userid" TYPE="int" LENGTH="10" NOTNULL="true" UNSIGNED="true" DEFAULT="0" SEQUENCE="false" PREVIOUS="cm" NEXT="identifier"/>
-        <FIELD NAME="identifier" TYPE="char" LENGTH="255" NOTNULL="true" SEQUENCE="false" PREVIOUS="userid" NEXT="timeretrieved"/>
-        <FIELD NAME="similarityscore" TYPE="int" LENGTH="5" NOTNULL="true" UNSIGNED="true" DEFAULT="0" SEQUENCE="false" PREVIOUS="statuscode" NEXT="attempt"/>
-	<FIELD NAME="timeretrieved" TYPE="int" LENGTH="10" NOTNULL="false" UNSIGNED="true" DEFAULT="0" SEQUENCE="false" PREVIOUS="identifier" NEXT="data"/>
+        <FIELD NAME="identifier" TYPE="char" LENGTH="255" NOTNULL="true" SEQUENCE="false" PREVIOUS="userid" NEXT="similarityscore"/>
+        <FIELD NAME="similarityscore" TYPE="int" LENGTH="5" NOTNULL="true" UNSIGNED="true" DEFAULT="0" SEQUENCE="false" PREVIOUS="identifier" NEXT="preliminary"/>
+        <FIELD NAME="preliminary" TYPE="int" LENGTH="2" NOTNULL="true" UNSIGNED="true" DEFAULT="0" SEQUENCE="false" PREVIOUS="similarityscore" NEXT="timeretrieved"/>
```
Comment (baholladay) [inline db/install.xml:14]: Make sure to uninstall and drop the tables and reinstall VC to test this code

File: lib.php [hunk 2]
```diff
@@ -206,24 +210,29 @@ public function get_file_results($cmid, $userid, $file, $vericite=null) {
...
 			//there should only be 1 result
             foreach ($contentscore as $content) {
                 $mycontent = $content;
-				if(isset($mycontent->timesubmitted) && (time() - $mycontent->timesubmitted) < (60 * PLAGIARISM_VERICITE_SCORE_CACHE_IGNORE_MIN)){
-					//don't consult the cache until at least PLAGIARISM_VERICITE_SCORE_CACHE_IGNORE_MIN after submission
-					plagiarism_vericite_log("don't consult cache");
-					//recheck after PLAGIARISM_VERICITE_SCORE_CACHE_IGNORE_MIN_RECHECK minutes instead of the normal cache time
-					$cacheTime = 60 * PLAGIARISM_VERICITE_SCORE_CACHE_IGNORE_MIN_RECHECK;
+
+				if(!isset($mycontent->preliminary) || $mycontent->preliminary==1){
```
Comment (baholladay) [inline lib.php]: I don't understand the "!isset" part of this condition

File: lib.php [hunk 3]
```diff
@@ -206,24 +210,29 @@ public function get_file_results($cmid, $userid, $file, $vericite=null) {
 		//do not consult the cache if the timesubmitted is less than PLAGIARISM_VERICITE_SCORE_CACHE_IGNORE_MIN
 		$cacheTime = 60 * PLAGIARISM_VERICITE_SCORE_CACHE_MIN;
         $mycontent = null;
-        $contentscore = $DB->get_records('plagiarism_vericite_files', array('cm' => $vericite['cmid'], 'userid' => $userid, 'identifier' => $fileid), '', 'id,cm,userid,identifier,similarityscore, timeretrieved, status, timesubmitted');
+        $contentscore = $DB->get_records('plagiarism_vericite_files', array('cm' => $vericite['cmid'], 'userid' => $userid, 'identifier' => $fileid), '', 'id,cm,userid,identifier,similarityscore, preliminary, timeretrieved, status, timesubmitted');
         if (!empty($contentscore)) {
 			//there should only be 1 result
             foreach ($contentscore as $content) {
                 $mycontent = $content;
-				if(isset($mycontent->timesubmitted) && (time() - $mycontent->timesubmitted) < (60 * PLAGIARISM_VERICITE_SCORE_CACHE_IGNORE_MIN)){
```
Comment (baholladay) [inline lib.php]: I noticed you removed the min recheck flag in these conditions. Have you watched the logs to make sure that we do not call our API too many times in cases with multiple submissions that are all outside of the cache limit? I believe that min recheck was a safe guard against that.

File: lib.php [hunk 4]
```diff
@@ -206,24 +210,29 @@ public function get_file_results($cmid, $userid, $file, $vericite=null) {
...
-					//don't consult the cache until at least PLAGIARISM_VERICITE_SCORE_CACHE_IGNORE_MIN after submission
-					plagiarism_vericite_log("don't consult cache");
-					//recheck after PLAGIARISM_VERICITE_SCORE_CACHE_IGNORE_MIN_RECHECK minutes instead of the normal cache time
-					$cacheTime = 60 * PLAGIARISM_VERICITE_SCORE_CACHE_IGNORE_MIN_RECHECK;
+
+				if(!isset($mycontent->preliminary) || $mycontent->preliminary==1){
+                    plagiarism_vericite_log("don't consult cache");
```
Comment (baholladay) [inline lib.php]: why do you log "don't consult cache, but you have a "cachetime" value > 0

File: lib.php [hunk 5]
```diff
@@ -798,6 +812,24 @@ function plagiarism_vericite_ends_with($str, $test) {
...
 
+/**
+ * Time expired when the difference between now and the last time we've checked it, is greater than the cacheTime
+ * @param $timeRetrieved
+ * @param $cacheTime
+ * @return bool
+ */
+function plagiarism_vericite_is_cache_time_expired($timeRetrieved, $cacheTime) {
+    plagiarism_vericite_log("now(): " . time() . " retrieved: " . $timeRetrieved . " cacheTime: " . $cacheTime . " timeLeft: " . plagiarism_vericite_time_left_in_cache($timeRetrieved, $cacheTime));
+    if($timeRetrieved == 0) {
```
Comment (baholladay) [inline lib.php]: why is this needed? If timeRetrieved is 0, then time() - 0 should be greater than the cacheTime

**Comment by baholladay** [comment]
Please try to undo all of the changes surrounding the cache timeout and make it as simple as possible like:

cacheTime = isPreliminary? PLAGIARISM_VERICITE_SCORE_CACHE_PRELIMINAYR_IGNORE_MIN : PLAGIARISM_VERICITE_SCORE_CACHE_IGNORE_MIN

File: sdk/lib/Model/ReportScoreReponse.php [hunk 6]
```diff
@@ -69,7 +69,9 @@ class ReportScoreReponse implements ArrayAccess
         'user' => 'string',
         'assignment' => 'string',
         'external_content_id' => 'string',
-        'score' => 'int'
+        'preliminary' => 'bool',
+        'score' => 'int',
+        'user' => 'string'
```
Comment (baholladay) [inline sdk/lib/Model/ReportScoreReponse.php]: user is here twice

File: sdk/lib/Model/ReportScoreReponse.php [hunk 7]
```diff
@@ -85,7 +87,9 @@ public static function swaggerTypes()
         'user' => 'user',
         'assignment' => 'assignment',
         'external_content_id' => 'externalContentId',
-        'score' => 'score'
+        'preliminary' => 'preliminary',
+        'score' => 'score',
+        'user' => 'user'
     );
```
Comment (baholladay) [inline sdk/lib/Model/ReportScoreReponse.php]: user is here twice

File: sdk/lib/Model/ReportScoreReponse.php [hunk 8]
```diff
@@ -101,7 +105,9 @@ public static function attributeMap()
         'user' => 'setUser',
         'assignment' => 'setAssignment',
         'external_content_id' => 'setExternalContentId',
-        'score' => 'setScore'
+        'preliminary' => 'setPreliminary',
+        'score' => 'setScore',
+        'user' => 'setUser'
     );
```
Comment (baholladay) [inline sdk/lib/Model/ReportScoreReponse.php]: user is in here twice

File: sdk/lib/Model/ReportScoreReponse.php [hunk 9]
```diff
@@ -117,7 +123,9 @@ public static function setters()
         'user' => 'getUser',
         'assignment' => 'getAssignment',
         'external_content_id' => 'getExternalContentId',
-        'score' => 'getScore'
+        'preliminary' => 'getPreliminary',
+        'score' => 'getScore',
+        'user' => 'getUser'
     );
```
Comment (baholladay) [inline sdk/lib/Model/ReportScoreReponse.php]: user is in here twice

**Comment by peterdietz** [comment]
Pushed updated PR that is smaller.
